perun-engine/src/main/java/cz/metacentrum/perun/engine/scheduling/impl/BlockingGenExecutorCompletionService.java

```diff
@@ -25,65 +23,107 @@
  * @see cz.metacentrum.perun.engine.runners.GenCollector
  *
  * @author David Šarman
+ * @author Pavel Zlámal <zlamal@cesnet.cz>
  */
 public class BlockingGenExecutorCompletionService implements BlockingCompletionService<Task> {
 
 	private final static Logger log = LoggerFactory.getLogger(BlockingGenExecutorCompletionService.class);
 	private CompletionService<Task> completionService;
-	@Autowired
-	@Qualifier("generatingTasks")
-	private BlockingBoundedMap<Integer, Task> executingTasks;
-	private int limit;
+	private ConcurrentMap<Future<Task>, Task> executingGenTasks = new ConcurrentHashMap<>();
+	private Semaphore semaphore;
 
 	/**
 	 * Create new blocking CompletionService for GEN Tasks with specified limit
 	 *
 	 * @param limit Limit for processing GEN Tasks
 	 */
 	public BlockingGenExecutorCompletionService(int limit) {
-		this.limit = limit;
 		completionService = new ExecutorCompletionService<Task>(Executors.newFixedThreadPool(limit), new LinkedBlockingQueue<Future<Task>>());
+		this.semaphore = new Semaphore(limit);
 	}
 
 	@Override
 	public Future<Task> blockingSubmit(EngineWorker<Task> taskWorker) throws InterruptedException {
+		semaphore.acquire();
```

Comment:
You should expect the following code in this method could throw an exception, in such case you might to release the semaphore.